### ISSUE TITLE
haxe.Http - JavaScript requesting local file not possible?

### ISSUE THREAD

**Issue by derRaab** [HaxeFoundation/haxe#2075]
A couple of weeks ago I started a discussion at the Google User Group https://groups.google.com/forum/?hl=en#!topic/haxelang/6OcWJey4Odg about the problems with local file access from JavaScript (since most of the browsers locally simulate a HTTP status of 0).

At the end I came up with a simple port of the jQuery solution to this basic issue which basically looks like this:

```javascript
haxe.Http.prototype = {
...
 ,request: function(post) {
  var me = this;
  var r = new js.XMLHttpRequest();
  var onreadystatechange = function() {
   if(r.readyState != 4) return;
    var s = (function($this) {
    var $r;
    try {
     $r = r.status;
    } catch( e ) {
     $r = null;
    }

    // local protocol fix borrowed from jQuery:


    // If the request is local and we have data: assume a success
    // (success with no data won't get notified, that's the best we
    // can do given current implementations)
    if (!$r) {
        $r = (function($r) {
            try {
                var ajaxLocation = location.href;
            } catch( e ) {
                ajaxLocation = document.createElement( "a" );
                ajaxLocation.href = "";
                ajaxLocation = ajaxLocation.href;
            }
            var rurl = /^([\w.+-]+:)(?:\/\/([^\/?#:]*)(?::(\d+)|)|)/,
                rlocalProtocol = /^(?:about|app|app-storage|.+-extension|file|res|widget):$/,
                ajaxLocParts = rurl.exec( ajaxLocation.toLowerCase() ) || [],
                isLocal = rlocalProtocol.test( ajaxLocParts[ 1 ] );
            if (isLocal) {
                $r = r.responseText ? 200 : 404;
            }
            return $r;
        })($r);
    }

   return $r;
  }(this));
```
BUT - I'm not happy with the implementation since it would be a much better approach to evaluate the local environment only once. Maybe it would be better to store it in js.isLocal or somewhere else?

Since I now start to develop local HTML / JS applications I would love to add this fix somewhere - but I'm not sure where to start? What shall I do next?

**Comment by aduros** [comment]
Switching milestone to long term. As a workaround during development, you can turn off sandboxing in your browser, or just use a web server.

**Comment by Justinfront** [comment]
derRaab
I don't think jQuery should be a requirement for Haxe JS it maybe batteries included but it can be excluded and probably should be removed from Haxe core in favour of haxelib, thus to propose something like this be internal to Haxe js is not a good idea, I am sure there are lighter better approaches to the general problem. It's very easy to use neko or nodejs server, but if your solution is suitably useful why not make it into a haxelib for others to use.
Best Justin

**Comment by derRaab** [comment]
Hi Justin, I think this issue was closed because it's an duplicate to https://github.com/HaxeFoundation/haxe/pull/3353  -> which is already merged into the master branch. And no - of course it's not an good idea to have jQuery as a requirement. I just use the code snippet from jQuery to enable local HTTP requests. :D I'm looking forward to 3.2!
